cpp/src/parquet/encryption/secure_string.cc

```diff
@@ -0,0 +1,111 @@
+#include "parquet/encryption/secure_string.h"
+
+#include <openssl/crypto.h>
+#include <openssl/opensslv.h>
+#include <utility>
+#if defined(_WIN32)
+#  include <windows.h>
+#endif
+
+#include "arrow/util/span.h"
+#include "parquet/encryption/encryption.h"
+
+namespace parquet::encryption {
+SecureString::SecureString(SecureString&& secret) noexcept
+    : secret_(std::move(secret.secret_)) {}
+SecureString::SecureString(std::string&& secret) noexcept : secret_(std::move(secret)) {
+  SecureClear(&secret);
+}
+
+SecureString& SecureString::operator=(SecureString&& secret) noexcept {
+  if (this == &secret) {
+    // self-assignment
+    return *this;
+  }
+  Dispose();
+  secret_ = std::move(secret.secret_);
+  return *this;
+}
+SecureString& SecureString::operator=(const SecureString& secret) noexcept {
+  if (this == &secret) {
+    // self-assignment
+    return *this;
+  }
+  Dispose();
+  secret_ = secret.secret_;
+  return *this;
+}
+SecureString& SecureString::operator=(std::string&& secret) noexcept {
+  Dispose();
+  secret_ = std::move(secret);
+  SecureClear(&secret);
+  return *this;
+}
+
+bool SecureString::operator==(const SecureString& other) const {
+  return secret_ == other.secret_;
+}
+
+bool SecureString::operator!=(const SecureString& other) const {
+  return secret_ != other.secret_;
+}
+
+bool SecureString::empty() const { return secret_.empty(); }
+std::size_t SecureString::size() const { return secret_.size(); }
+std::size_t SecureString::length() const { return secret_.length(); }
+
+::arrow::util::span<uint8_t> SecureString::as_span() {
+  return {reinterpret_cast<uint8_t*>(secret_.data()), secret_.size()};
+}
+::arrow::util::span<const uint8_t> SecureString::as_span() const {
+  return {reinterpret_cast<const uint8_t*>(secret_.data()), secret_.size()};
+}
+std::string_view SecureString::as_view() const {
+  return {secret_.data(), secret_.size()};
+}
+
+void SecureString::Dispose() { SecureClear(&secret_); }
+void SecureString::SecureClear(std::string* secret) {
+  secret->clear();
+  SecureClear(reinterpret_cast<uint8_t*>(secret->data()), secret->capacity());
+}
+inline void SecureString::SecureClear(uint8_t* data, size_t size) {
+  // Heavily borrowed from libb2's `secure_zero_memory` at
+  // https://github.com/BLAKE2/libb2/blob/master/src/blake2-impl.h
+#if defined(_WIN32)
+  SecureZeroMemory(data, size);
+#elif defined(__STDC_LIB_EXT1__)
+  // memset_s is meant to not be optimized away
+  memset_s(data, size, 0, size);
+#elif defined(OPENSSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x30000000
+  OPENSSL_cleanse(data, size);
+#elif defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ >= 25))
+  // glibc 2.25+ has explicit_bzero
+  explicit_bzero(data, size);
+#else
+  // Try to ensure that a true library call to memset() will be generated
+  // by the compiler.
+  static const volatile auto memset_v = &memset;
+  memset_v(data, 0, size);
+  __asm__ __volatile__("" ::"r"(data) : "memory");
```

Comment:
It seems questionable to me that the ultimate fallback should not be a simple loop

```suggestion
  for (size_t i = 0; i < size; ++i) {
    data[i] = 0;
  }
```

I don't know how cross platform this `__asm__` block is.


Also, since this code is borrowed from another project it needs to be mentioned in LICENSE.txt, for example https://github.com/apache/arrow/blob/7df396eec4c620ddc4a4db780ea0d5ffa4cf685b/LICENSE.txt#L1541-L1543

Reply:
What about
```c++
  volatile char *p = data;
  while (size--) *p++ = 0;
```
https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://cplusplus.com/articles/ETqpX9L8/&ved=2ahUKEwjF5MvK-8-MAxUFnf0HHUiTAcYQFnoECBEQAQ&usg=AOvVaw2otjQoJ6qGsoLpmq4cIidL

Reply:
Given https://github.com/BLAKE2/libb2/blob/master/src/blake2-impl.h is [CC0 Public Domain Dedication](http://creativecommons.org/publicdomain/zero/1.0/), and our code deviates a lot, you could say it is as much inspired by that implementation than it is by all those variations available on stackoverflow.

I don't think this qualifies as a copy or derived work.

Reply:
If it's in the public domain, then I imagine an inline comment referencing that should be sufficient. @pitrou do you agree? We do also account for some public domain artifacts [in LICENSE.txt](https://github.com/apache/arrow/blob/7df396eec4c620ddc4a4db780ea0d5ffa4cf685b/LICENSE.txt#L1466-L1470)

Reply:
We could add a comment, yes.

Reply:
@pitrou @bkietz

I have added more comments and links to prior art. I have also wrapped the `__asm__` to enable it only for GCC and Clang: https://github.com/apache/arrow/pull/46017/commits/510349c170834e348305a70d51face81a5709196

We have two options for the fallback branch:
1. the current `memset` implementation
2. the `while` loop below

Both benefit from the `__asm__ __volatile__` call.

```c++
  // Volatile pointer to the data is an attempt to avoid
  // that the compiler optimizes away the while loop.
  // https://cplusplus.com/articles/ETqpX9L8/
  // http://stackoverflow.com/a/13299459/965097
  // https://www.cryptologie.net/article/419/zeroing-memory-compiler-optimizations-and-memset_s/
  volatile char *p = data;
  while (size--) *p++ = 0;
```